pkg/services/chgm/chgm.go

```diff
@@ -50,32 +58,117 @@ type Service interface {
-// to create 'Client' you can use a mock, or fill it with
-// this differs from the ChgmProvider as it makes sure you are intentional when using functions.
-// if I am missing a func I will copy it from the corresponding package to the interface instead of
-// having a function change break my code.
-// TODO: decide if the Client should be the ChgmProvider
+// Client for the chgm investigation
 type Client struct {
 	Service
-	cluster *v1.Cluster
-	cd      *hivev1.ClusterDeployment
+	Cluster           *v1.Cluster
+	ClusterDeployment *hivev1.ClusterDeployment
+	*investigation.Client
+}
+
+// Triggered will analyse chgm incidents that are newly created
+func (c *Client) Triggered() error {
+	res, err := c.investigateStoppedInstances()
+	if err != nil {
+		return c.UpdateAndEscalateAlert(fmt.Sprintf("InvestigateInstances failed: %s\n", err.Error()))
+	}
+
+	fmt.Printf("the investigation returned %#v\n", res)
+
+	lsExists, err := c.LimitedSupportReasonsExist(c.Cluster.ID())
+	if err != nil {
+		return c.UpdateAndEscalateAlert(fmt.Errorf("failed to determine if limited support reason already exists: %w", err).Error())
+	}
+
+	// if lsExists, silence alert and add investigation to notes
+	if lsExists {
+		fmt.Println("Unrelated limited support reason present on cluster, silencing")
+		return c.SilenceAlert(res.String() + "Unrelated limited support reason present on cluster, silenced.")
+	}
+
+	if res.UserAuthorized {
+		fmt.Println("The node shutdown was not the customer. Should alert SRE")
+		return c.UpdateAndEscalateAlert(res.String())
+	}
+
+	// The node shutdown was the customer
+	// Put into limited support, silence and update incident notes
+	fmt.Printf("Sending limited support reason: %s\n", chgmLimitedSupport.Summary)
+	return utils.Retry(3, time.Second*2, func() error {
+		return c.PostLimitedSupport(res.String())
+	})
+}
+
+// Resolved will take appropriate action against a cluster whose CHGM incident has resolved.
+func (c *Client) Resolved() error {
+	res, err := c.investigateStartedInstances()
+
+	// The investigation encountered an error & never completed - alert Primary and report the error, retrying as many times as necessary to escalate the issue
+	if err != nil {
+		err = c.AddNote(fmt.Sprintf("Resolved investigation did not complete: %v\n", err.Error()))
+		if err != nil {
+			fmt.Println("could not update incident notes")
+		}
+		return utils.Retry(3, time.Second*2, func() error {
+			return c.CreateNewAlert(c.getInvestigationFailureAlert(err), c.GetServiceID())
+		})
+	}
+
+	// Investigation completed, but the state in OCM indicated the cluster didn't need investigation
+	if res.ClusterNotEvaluated {
+		fmt.Printf("Cluster has state '%s' in OCM, and so investigation is not need\n", res.ClusterState)
+		err = utils.Retry(3, time.Second*2, func() error {
+			return c.AddNote(res.String())
+		})
+		if err != nil {
+			fmt.Printf("Failed to add notes to incident: %s\n", res.String())
+		}
```

Comment:
**issue:** We used to `return nil` in this case, now we don't. This means that if a cluster doesn't need an investigation, we use the result of the investigation anyway and move forward with the logic after this.

Reply:
good catch! that got lost when i tried to untangle the exit paths